miriway.cpp

```diff
@@ -99,6 +119,42 @@ struct ShellComponents
     std::vector<std::vector<std::string>> commands;
 };
 
+class LockScreen
+{
+public:
+    explicit LockScreen(
+        std::function<void(std::vector<std::string> const& command_line)> launch,
+        std::function<void(std::string name)> const& conditionally_enable)
+        : lockscreen_option(
+            [&](mir::optional_value<std::string> const& app) {
+                    if (app.is_set())
+                        lockscreen_app = app.value();
+                },
+                "lockscreen-app",
+                "Lockscreen app to be triggered when the compositor session is locked"
+            ),
+            session_locker(
+              [&]{
+                  if (lockscreen_app) launch(ExternalClientLauncher::split_command(*lockscreen_app));
+              },
+              [] {}
+          )
+    {
+        conditionally_enable(WaylandExtensions::ext_session_lock_manager_v1);
+    }
+
+    void operator()(mir::Server& server) const
+    {
+        lockscreen_option(server);
+        session_locker(server);
+    }
+
+private:
+    ConfigurationOption const lockscreen_option;
+    SessionLockListener const session_locker;
+    std::optional<std::string> lockscreen_app;
+};
```

Comment:
Several things:

1. Capturing constructor args _by reference_ won't work for the SessionLocker callback
2. Please use uniform initialization syntax

I shuffled the code a little to reduce the #ifs:

```suggestion
class LockScreen
{
public:
    explicit LockScreen(
        std::function<void(std::vector<std::string> const& command_line)> launch,
        std::function<void(std::string name)> const& conditionally_enable) : 
            launch{std::move(launch)},
            lockscreen_option{[this](mir::optional_value<std::string> const& app) {
                    if (app.is_set())
                    {
#if MIRAL_VERSION < MIR_VERSION_NUMBER(5, 0, 0)
                        mir::log_warning("SessionLockListener is not available before miral 5.0");
#endif
                        lockscreen_app = app.value();
                    }
                },
                "lockscreen-app",
                "Lockscreen app to be triggered when the compositor session is locked"
          }
    {
        conditionally_enable(WaylandExtensions::ext_session_lock_manager_v1);
    }

    void operator()(mir::Server& server) const
    {
        lockscreen_option(server);
        session_locker(server);
    }

private:
    std::function<void(std::vector<std::string> const& command_line)> const launch;
    ConfigurationOption const lockscreen_option;
#if MIRAL_VERSION >= MIR_VERSION_NUMBER(5, 0, 0)
    SessionLockListener const session_locker{
              [this]{ if (lockscreen_app) launch(ExternalClientLauncher::split_command(*lockscreen_app)); },
              [] {}
          };
#eise
      auto const session_locker = [](mir::Server&){};
#endif
    std::optional<std::string> lockscreen_app;
};
#endif
```